File: parser/src/contract.ts
import { AlpObject } from './reader';

export interface ContractViolation {
  contractId: string;
  rule: string;
  reason: string;
  context: Record<string, unknown>;
}

export interface ContractResult {
  ok: boolean;
  violation?: ContractViolation;
}

export interface ContractObject {
  id: string;
  name?: string;
  from: string;
  to: string;
  type: 'api' | 'data' | 'tool' | 'repo';
  requires: string[];
  allows: string[];
  denies: string[];
  on_violation: 'deny' | 'warn' | 'log';
}

export class ContractEngine {
  private contracts: Map<string, ContractObject>;

  constructor(objects: AlpObject[]) {
    this.contracts = new Map();
    for (const obj of objects) {
      if (obj._type === 'contract') {
        this.contracts.set(obj.id as string, normalize(obj));
      }
    }
  }

  get count(): number {
    return this.contracts.size;
  }

  /**
   * Check whether `context` satisfies the contract with the given id.
   *
   * @param contractId The `@contract` id to evaluate.
   * @param context    Must include `operation: string`. May include arbitrary
   *                   fields referenced by `requires` expressions.
   */
  check(contractId: string, context: Record<string, unknown>): ContractResult {
    const contract = this.contracts.get(contractId);
    if (!contract) {
      return {
        ok: false,
        violation: { contractId, rule: '', reason: `contract '${contractId}' not found`, context },
      };
    }

    for (const req of contract.requires) {
      if (!evaluateRequire(req, context)) {
        return this.violation(contract, req, 'required condition not met', context);
      }
    }

    const operation = String(context.operation ?? '');

    if (contract.denies.some((d) => matchesGlob(operation, d))) {
      return this.violation(contract, operation, 'denied', context);
    }

    if (contract.allows.length > 0 && !contract.allows.includes(operation)) {
      return this.violation(contract, operation, 'not in allow-list', context);
    }

    return { ok: true };
  }

  /** List all contracts. */
  list(): ContractObject[] {
    return Array.from(this.contracts.values());
  }

  private violation(contract: ContractObject, rule: string, reason: string, context: Record<string, unknown>): ContractResult {
    const v: ContractViolation = { contractId: contract.id, rule, reason, context };
    if (contract.on_violation === 'log') {
      console.log(`[contract] violation: ${contract.id} — ${rule}: ${reason}`);
    }
    if (contract.on_violation === 'warn') {
      console.warn(`[contract] violation (warn): ${contract.id} — ${rule}: ${reason}`);
      return { ok: true };
    }
    return { ok: false, violation: v };
  }
}

function normalize(obj: AlpObject): ContractObject {
  return {
    id: obj.id as string,
    name: obj.name as string | undefined,
    from: obj.from as string,
    to: obj.to as string,
    type: (obj.type as ContractObject['type']) ?? 'api',
    requires: Array.isArray(obj.requires) ? obj.requires.map(String) : [],
    allows: Array.isArray(obj.allows) ? obj.allows.map(String) : [],
    denies: Array.isArray(obj.denies) ? obj.denies.map(String) : [],
    on_violation: (obj.on_violation as ContractObject['on_violation']) ?? 'deny',
  };
}

function evaluateRequire(expr: string, context: Record<string, unknown>): boolean {
  const trimmed = expr.trim();
  for (const op of ['<=', '>=', '!=', '==', '<', '>']) {
    if (trimmed.includes(op)) {
      const idx = trimmed.indexOf(op);
      const key = trimmed.slice(0, idx).trim();
      const rawValue = trimmed.slice(idx + op.length).trim();
      const actual = getNested(context, key);
      const expected = parseValue(rawValue);
      if (actual == null) return false;
      switch (op) {
        case '<':  return Number(actual) < Number(expected);
        case '>':  return Number(actual) > Number(expected);
        case '<=': return Number(actual) <= Number(expected);
        case '>=': return Number(actual) >= Number(expected);
        case '==': return actual === expected;
        case '!=': return actual !== expected;
      }
    }
  }
  const parts = trimmed.split(/\s+/);
  if (parts.length === 2) {
    const actual = getNested(context, parts[0]);
    const expected = parseValue(parts[1]);
    return actual === expected;
  }
  return getNested(context, trimmed) != null && getNested(context, trimmed) !== false;
}

function getNested(context: Record<string, unknown>, key: string): unknown {
  const parts = key.split('.');
  let cur: unknown = context;
  for (const part of parts) {
    if (typeof cur === 'object' && cur !== null && part in (cur as Record<string, unknown>)) {
      cur = (cur as Record<string, unknown>)[part];
    } else {
      return undefined;
    }
  }
  return cur;
}

function parseValue(raw: string): unknown {
  if (raw === 'true') return true;
  if (raw === 'false') return false;
  const n = Number(raw);
  if (!isNaN(n)) return n;
  return raw;
}

function matchesGlob(value: string, pattern: string): boolean {
  if (pattern.endsWith('.*')) {
    const prefix = pattern.slice(0, -2);
    return value.startsWith(prefix);
  }
  return value === pattern;
}

File: sdk/python/alp_sdk/contract.py
"""ALP contract engine (v8.3.0 - Python SDK parity).

Mirrors the TypeScript ``ContractEngine``: validates whether a handoff
context satisfies a ``@contract`` object's ``requires``, ``allows``, and
``denies`` rules. Returns a ``ContractResult`` indicating pass/fail and an
optional ``ContractViolation``.
"""
from __future__ import annotations


from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Union

from .models import AlpObject


@dataclass
class ContractViolation:
    contract_id: str
    rule: str
    reason: str
    context: Dict[str, Any] = field(default_factory=dict)


@dataclass
class ContractResult:
    ok: bool
    violation: Optional[ContractViolation] = None


@dataclass
class ContractObject:
    id: str
    name: Optional[str] = None
    from_: str = ''
    to: str = ''
    type: str = 'api'
    requires: List[str] = field(default_factory=list)
    allows: List[str] = field(default_factory=list)
    denies: List[str] = field(default_factory=list)
    on_violation: str = 'deny'


class ContractEngine:
    """Evaluate ``@contract`` objects against handoff contexts."""

    def __init__(self, objects: List[AlpObject]):
        self.contracts: Dict[str, ContractObject] = {}
        for obj in objects:
            if obj._type == 'contract':
                c = self._normalize(obj)
                self.contracts[c.id] = c

    @property
    def count(self) -> int:
        return len(self.contracts)

    def check(self, contract_id: str, context: Dict[str, Any]) -> ContractResult:
        contract = self.contracts.get(contract_id)
        if contract is None:
            return ContractResult(
                ok=False,
                violation=ContractViolation(
                    contract_id=contract_id,
                    rule='',
                    reason=f"contract '{contract_id}' not found",
                    context=context,
                ),
            )

        for req in contract.requires:
            if not _evaluate_require(req, context):
                return self._violation(contract, req, 'required condition not met', context)

        operation = str(context.get('operation', ''))

        if any(_matches_glob(operation, d) for d in contract.denies):
            return self._violation(contract, operation, 'denied', context)

        if contract.allows and operation not in contract.allows:
            return self._violation(contract, operation, 'not in allow-list', context)

        return ContractResult(ok=True)

    def list(self) -> List[ContractObject]:
        return list(self.contracts.values())

    def _violation(self, contract: ContractObject, rule: str, reason: str, context: Dict[str, Any]) -> ContractResult:
        violation = ContractViolation(contract_id=contract.id, rule=rule, reason=reason, context=context)
        if contract.on_violation == 'log':
            print(f"[contract] violation: {contract.id} — {rule}: {reason}")
        if contract.on_violation == 'warn':
            print(f"[contract] violation (warn): {contract.id} — {rule}: {reason}")
            return ContractResult(ok=True)
        return ContractResult(ok=False, violation=violation)

    @staticmethod
    def _normalize(obj: AlpObject) -> ContractObject:
        return ContractObject(
            id=obj.properties.get('id', ''),
            name=obj.properties.get('name'),
            from_=obj.properties.get('from', ''),
            to=obj.properties.get('to', ''),
            type=obj.properties.get('type', 'api'),
            requires=[str(r) for r in (obj.properties.get('requires') or [])],
            allows=[str(r) for r in (obj.properties.get('allows') or [])],
            denies=[str(r) for r in (obj.properties.get('denies') or [])],
            on_violation=obj.properties.get('on_violation', 'deny'),
        )


def _evaluate_require(expr: str, context: Dict[str, Any]) -> bool:
    expr = expr.strip()
    for op in ('<=', '>=', '!=', '==', '<', '>'):
        if op in expr:
            parts = expr.split(op, 1)
            if len(parts) == 2:
                key = parts[0].strip()
                raw_value = parts[1].strip()
                actual = _get_nested(context, key)
                expected = _parse_value(raw_value)
                if actual is None:
                    return False
                try:
                    if op == '<':  return float(actual) < float(expected)
                    if op == '>':  return float(actual) > float(expected)
                    if op == '<=': return float(actual) <= float(expected)
                    if op == '>=': return float(actual) >= float(expected)
                    if op == '==': return actual == expected
                    if op == '!=': return actual != expected
                except (TypeError, ValueError):
                    return False
            return True
    # No comparison operator — try "key value" as implicit equality.
    parts = expr.split()
    if len(parts) == 2:
        key, raw_value = parts
        actual = _get_nested(context, key)
        expected = _parse_value(raw_value)
        return actual == expected
    val = _get_nested(context, expr)
    return val is not None and val is not False


def _get_nested(context: Dict[str, Any], key: str) -> Any:
    parts = key.split('.')
    cur: Any = context
    for part in parts:
        if isinstance(cur, dict) and part in cur:
            cur = cur[part]
        else:
            return None
    return cur


def _parse_value(raw: str) -> Any:
    if raw == 'true':
        return True
    if raw == 'false':
        return False
    try:
        return int(raw)
    except ValueError:
        pass
    try:
        return float(raw)
    except ValueError:
        pass
    return raw


def _matches_glob(value: str, pattern: str) -> bool:
    if pattern.endswith('.*'):
        prefix = pattern[:-2]
        return value.startswith(prefix)
    return value == pattern
